Upgrade `ruff` and replace `black` with `ruff format`

jsma opened the pull request:
The vast majority of the changes introduced by `ruff format` is to remove empty lines and remove unnecessary `()` wrappers around variables and multi-line statements that did not need to be multi-line statements.

I only manually edited the search query files to use f-strings in the error messages to make them one-liners because `ruff format` moved the second line of the string up to the first line, but did not merge them. For example, `ruff` turned this:

```python
"Lexeme can only be combined with other Lexemes, "
                "got {}.".format(type(other))
```

Into this:
```python
"Lexeme can only be combined with other Lexemes, " "got {}.".format(type(other))
```

_Please check the following:_

-   [x] Do the tests still pass?[^1]
-   [x] Does the code comply with the style guide?
    -   [x] Run `make lint` from the Wagtail root.

jsma commented:
It looks like the failures are all related to an outage with the [Docker registry](https://www.dockerstatus.com/):

<img width="964" alt="Screenshot 2023-11-11 at 9 28 16 AM" src="https://github.com/wagtail/wagtail/assets/24574/3401d291-4850-4ff9-9b69-6a74346f8409">

lb- reviewed (commented):
I will let a few other core team members chime in but overall these changes look better to me, especially the removal of `()` when not required.

I am all for this change.

laymonage reviewed (commented):
FWIW I was planning to do this in the next few weeks, so +1 from me. Thanks!

jsma commented:
The Docker registry is now "all systems operational". Is a manual step required to retrigger the CI?

lb- commented:
@jsma just force push again. Maybe with a small change to the commit message or rebase on main if needed.

zerolab commented on .circleci/config.yml:
```diff
@@ -20,7 +20,6 @@ jobs:
           paths:
             - .venv
       - run: pipenv run ruff check .
-      - run: pipenv run black --target-version py38 --check --diff .
```
This should be `ruff format --check .` since all linting is happening in CircleCI

zerolab commented on Makefile:
```diff
@@ -18,7 +18,6 @@ develop: clean-pyc
 
 lint-server:
-	black --target-version py38 --check --diff .
 	ruff check .
```
```suggestion
	ruff format --check .
	ruff check .
```

zerolab reviewed (commented):
Thank you for this @jsma.
The changes LGTM. Left a couple of suggestions.

Normally I'd recommend doing two commits for this: first the deps/config changes, then a separate linting one. This is particularly useful when loads of files are changed, and if you want to use `.git-blame-ignore-revs`. But this is small enough that it is fine without, imho

laymonage commented on wagtail/search/backends/database/sqlite/query.py:
```diff
@@ -272,9 +269,7 @@ def normalize(search_query: SearchQuery) -> Tuple[SearchQuery]:
-        for (
-            subquery
-        ) in (
+        for subquery in (
             negated_subqueries
         ):  # If there's a MatchAll subquery, then anything will get matched.
```
```suggestion
        for subquery in negated_subqueries:
            # If there's a MatchAll subquery, then anything will get matched.
```

jsma replied:
Ah, good catch. It looks like `ruff format` is not formatting these lines due to the trailing comment. I'll move the comments to a newline and it should format these into single line statements.

laymonage commented on wagtail/search/backends/database/sqlite/query.py:
```diff
@@ -248,9 +247,7 @@ def normalize(search_query: SearchQuery) -> Tuple[SearchQuery]:
-        for (
-            subquery
-        ) in (
+        for subquery in (
             negated_subqueries
         ):  # If there's a negated MatchAll subquery, then nothing will get matched.
             if isinstance(subquery, MatchAll):
```
```suggestion
        for subquery in negated_subqueries:
            # If there's a negated MatchAll subquery, then nothing will get matched.
            if isinstance(subquery, MatchAll):
```

laymonage commented on wagtail/search/backends/database/mysql/mysql.py:
```diff
@@ -491,9 +491,7 @@ def search(self, config, start, stop, score_field=None):
-            if (
-                self.order_by_relevance
-            ):  # Only applies to the case where the outermost query is not a Not(), because if it is, the relevance score is always 0 (anything that matches is excluded from the results).
+            if self.order_by_relevance:  # Only applies to the case where the outermost query is not a Not(), because if it is, the relevance score is always 0 (anything that matches is excluded from the results).
```
```suggestion
            if self.order_by_relevance:
                # Only applies to the case where the outermost query is not a Not(),
                # because if it is, the relevance score is always 0
                # (anything that matches is excluded from the results).
```

laymonage commented on wagtail/models/audit_log.py:
```diff
@@ -69,9 +69,7 @@ def with_instances(self):
-        instances_by_id = (
-            {}
-        )  # lookup of (content_type_id, stringified_object_id) to instance
+        instances_by_id = {}  # lookup of (content_type_id, stringified_object_id) to instance
```
```suggestion
        # lookup of (content_type_id, stringified_object_id) to instance
        instances_by_id = {}
```

laymonage commented on wagtail/admin/models.py:
```diff
@@ -16,9 +16,7 @@
-        default_permissions = (
-            []
-        )  # don't create the default add / change / delete / view perms
+        default_permissions = []  # don't create the default add / change / delete / view perms
```
```suggestion
        # don't create the default add / change / delete / view perms
        default_permissions = []
```

laymonage commented on wagtail/blocks/struct_block.py:
```diff
@@ -155,9 +155,7 @@ def value_omitted_from_data(self, data, files, prefix):
-        result = (
-            []
-        )  # build up a list of (name, value) tuples to be passed to the StructValue constructor
+        result = []  # build up a list of (name, value) tuples to be passed to the StructValue constructor
```
```suggestion
        # build up a list of (name, value) tuples to be passed to the StructValue constructor
        result = []
```

jsma commented on wagtail/search/backends/database/sqlite/query.py:
```diff
@@ -248,9 +247,7 @@ def normalize(search_query: SearchQuery) -> Tuple[SearchQuery]:
         ):  # If there are no negated subqueries, return an And(), now without the redundant MatchAll subqueries.
```
@laymonage this file is full of trailing comments on multi-line statements that could be single line statements if the comments are moved to their own line. Do you want every instance of this changed?

laymonage replied:
Ah, OK. I wasn't sure which bits were manually modified and which ones were auto-formatted. The changes in `wagtail/search/backends/database/mysql/query.py` seem similar, hence I made the suggestion to do it here as well. If those changes were manually made, I think it's either we make them all, or undo those changes.

(The `.format()` issue in the PR description makes sense to be manually edited I think.)

https://github.com/wagtail/wagtail/pull/11220/files#diff-72a2e9508512a738e0447971a82c79c9920871465bba681913a08616bcecd29eR237-R240

jsma replied:
I don't recall what happened in that specific case and it looks like there is exactly one other instance where I may have manually moved the comments to get automated one-liner formatting. I could revert those two cases to see if `ruff format` had done something odd that I reacted to, or if I just go carried away while reviewing ;) Everything else was automated aside from fixing the odd string concatenation mentioned in my original post.

allcaps reviewed (changes requested):
Can we have two commits? One adding ruff command, and the other with reformatting of Python files?

The reformat commit hash should be added to git blame ignore refs (third commit).

https://github.com/wagtail/wagtail/blob/main/.git-blame-ignore-revs

jsma commented:
> Can we have two commits? One adding ruff command, and the other with reformatting of Python files?
> 
> The reformat commit hash should be added to git blame ignore refs (third commit).
> 
> https://github.com/wagtail/wagtail/blob/main/.git-blame-ignore-revs

We _could_. @zerolab [said](https://github.com/wagtail/wagtail/pull/11220#pullrequestreview-1726358685):

> Normally I'd recommend doing two commits for this: first the deps/config changes, then a separate linting one. This is particularly useful when loads of files are changed, and if you want to use .git-blame-ignore-revs. But this is small enough that it is fine without, imho

I'm awaiting feedback from @laymonage to see if I'm on the hook to do a bunch of manual formatting that `ruff format` did not touch due to trailing comments. If so, yes, this would be a much bigger job and a noisier commit touching more lines of code. 

The original commit mostly just removed blank lines and would not complicate the git blame history for lines of actual code. Future viewers of git blame would not even see the affected lines since they were removed, they would just see the blame entries for the actual code above and below the removed lines.

jsma commented on wagtail/search/backends/database/mysql/query.py:
```diff
@@ -235,10 +234,13 @@ def __init__(
-            "title",
-            "body",
-        ]  # We need to provide a default list of columns if the user doesn't specify one. We have a joint index for for 'title' and 'body' (see wagtail.search.migrations.0006_customise_indexentry), so we'll pick that one.
+        self.columns = (
```
@laymonage now I remember why I manually formatted this previously. With the trailing comment it wraps it all in `()` and adds to the multi-line goofiness when this could easily be a one-liner if the comment was placed above the code rather than inline. In the interests of getting this merged, I'll keep my current PR focused on the automatic `ruff format` changes and leave the manual cleanup for another day.

jsma commented:
I've split this into three commits:

1. Introduce the new version of `ruff` with the related configuration updates
2. Apply `ruff format .`
3. Git blame ignore the previous commit

Currently the only manual formatting is the odd string concatenation bits I pointed out in my original post above. Everything else is exactly as `ruff format .` automatically formatted it. Unless there's additional feedback, I believe I've addressed the concerns above and this is ready for final review. Thanks!

laymonage reviewed (approved):
I think this looks good, we can address the inline comments issue separately. Just a note to whoever's going to merge this, make sure to rebase and merge this with fast-forward (locally) and double check that the SHA in `.git-blame-ignore-revs` matches the actual commit in `main`. (Using GitHub's merge button will change the SHAs.)

laymonage commented:
Merged in 364c1cee0dba9a91a2811e9ef48a7f6093a1a91c (I can't force-push to your branch because it's on an organization account), thanks @jsma & others!
